19/06/24 workday integration

daniel-sussman opened the pull request.

daniel-sussman committed:
company_details for Workday

daniel-sussman committed:
Categorizer::CompanyDetailsFromUrl

daniel-sussman committed:
added methods to create subsidiaries for Workday companies

daniel-sussman committed:
Merge branch 'main' of github.com:Ches-ctrl/Cheddar into 19/06/24_Workday_integration

daniel-sussman committed:
makes initial request to jobs endpoint for facets json

daniel-sussman committed:
refactored facet search method

daniel-sussman committed:
can add jobs with Workday

daniel-sussman committed:
deleted some cassettes

daniel-sussman committed:
added parsing method for Workday

daniel-sussman committed:
added tests for Workday

daniel-sussman committed:
Delete spec/fixtures/cassettes/Company/Validations/1_2_2.yml

daniel-sussman committed:
updated workflow yaml

daniel-sussman committed:
Create 1_2_2.yml

daniel-sussman committed:
Merge branch '19/06/24_Workday_integration' of github.com:Ches-ctrl/Cheddar into 19/06/24_Workday_integration

daniel-sussman committed:
updated workflow yaml again

daniel-sussman committed:
adjusted parsing methods

daniel-sussman committed:
trying to fix tests not passing on GitHub...

daniel-sussman committed:
deleted Gemfile.lock

daniel-sussman committed:
trying to debug remotely

daniel-sussman committed:
removed the debugging line

daniel-sussman committed:
trying one more time...

daniel-sussman committed:
debugging remotely...

daniel-sussman committed:
removed gem caching

daniel-sussman committed:
cleared git cache

daniel-sussman committed:
debugging

daniel-sussman committed:
deleted and restored all VCR cassettes

daniel-sussman committed:
VCR config record :new_episodes

patanj101 commented on app/models/company.rb:
The methods create_all_relevant_jobs and fetch_description don't belong
directly in the Company model. Here's a breakdown and suggestions for
improvement:

*Issues:*

   - *Separation of Concerns:* These methods deal with interacting with the
   Applicant Tracking System (ATS) and fetching data, which is not directly
   related to managing company data within the model itself. They violate the
   Single Responsibility Principle by introducing logic unrelated to company
   attributes.
   - *Testability:* Testing these methods within the model can be
   cumbersome due to their external dependencies (ATS interaction).

*Improved Approach:*

   1. *Move to Service Objects: *Create separate service objects (classes)
   to encapsulate the logic for:


   - Fetching company jobs (`CompanyJobsFetcher`)
   - Fetching company description (`CompanyDescriptionFetcher`)

patanj101 replied:
```
class CompanyJobsFetcher
  def initialize(company)
    @company = company
  end

  def call
    # Logic to fetch jobs from the ATS using company data
    # ...
  end
end

class CompanyDescriptionFetcher
  def initialize(company)
    @company = company
  end

  def call
    # Logic to fetch description using website URL or other methods
    # ...
  end
end
```

patanj101 replied:
Follow the same structure, initiliaze, call, process, processable ... and then spefic methods.

```
# frozen_string_literal: true

class OpportunitiesFetcher < ApplicationTask
  def initialize(opportunities, params)
    @opportunities = opportunities
    @params = params
  end

  def call
    return false unless processable

    process
  end

  private

  def processable
    true
  end

  def process
    opportunities = OpportunitiesQuery.call(wanted_opportunities)
    apply_filters(opportunities)
  end

  def apply_filters(opportunities)
```

patanj101 replied:
Much better :-)

def create_all_relevant_jobs should be totally removed from Company model though.
fetch_description also ... A little bit more complex because of the callback.

But avoid callback as much as you can.

https://medium.com/@distillerytech/rails-callback-hell-and-pipeline-heaven-lets-talk-about-the-problem-d6cd6dc02d05

daniel-sussman committed:
refactorisé selon les indications de jb

daniel-sussman committed:
resolved merge issues

daniel-sussman committed:
#create_all_company_jobs is no more.

daniel-sussman committed:
tests passing locally

daniel-sussman committed:
debugging remotely

daniel-sussman committed:
made cassette filepath relative

daniel-sussman committed:
created VCR.log

daniel-sussman committed:
workflow file changed

daniel-sussman committed:
addressed VCR issue

daniel-sussman committed:
updated VCR cassettes

daniel-sussman commented:
JB, I've made the changes you suggested to Company.rb

There's now a CompanyJobsFetcher class that replaces #create_all_relevant_jobs, and I've adjusted the test suite accordingly.

No more :after_save callbacks in Company. Moved that workflow into CompanyCreator. Is that the kind of fix you intended, or do you think we need something like the Pipelines class in the article you linked?

daniel-sussman committed:
changed call method for CompanyDescriptionFetcher

Ches-ctrl commented:
Notes:
- Calling crunchbase API that will update details (mutable data) - will need to reference data pipeline going forward
- Couple of updates on new...call can be made
- More splitting out into PORO objects possible
- Workday will need to be put into CompanyCreator format following this
